File: backend/budget-service/src/main/java/com/ohchurus/budget/service/CategoryService.java
package com.ohchurus.budget.service;

import com.ohchurus.budget.dto.input.CategoryFilterDTO;
import com.ohchurus.budget.dto.input.CategorySaveDTO;
import com.ohchurus.budget.dto.output.ResultDTO;

public interface CategoryService {

    ResultDTO saveAndUpdate(CategorySaveDTO dto);

    ResultDTO getById(Long id);

    ResultDTO getAll(CategoryFilterDTO filter);

    ResultDTO getTree(Long userId);

    ResultDTO delete(Long id);

    ResultDTO typeList();
}

File: frontend/src/services/categoryService.ts
import api from './api';
import { ResultDTO, Category, CategoryTree, CategoryFilter, PageDTO } from '../types';

const BASE = '/BUDGET-SERVICE/oh-churus/v1/categories';

export const categoryService = {
  save: async (data: Partial<Category>) => {
    const response = await api.post<ResultDTO<Category>>(
      `${BASE}/save`,
      data,
    );
    return response.data;
  },

  getById: async (id: string) => {
    const response = await api.post<ResultDTO<Category>>(
      `${BASE}/get/${id}`,
    );
    return response.data;
  },

  getAll: async (filter: CategoryFilter) => {
    const response = await api.post<ResultDTO<PageDTO<Category>>>(
      `${BASE}/all`,
      filter,
    );
    return response.data;
  },

  getTree: async (userId: string) => {
    const response = await api.post<ResultDTO<CategoryTree[]>>(
      `${BASE}/tree`,
      { userId },
    );
    return response.data;
  },

  delete: async (id: string) => {
    const response = await api.post<ResultDTO<null>>(
      `${BASE}/delete/${id}`,
    );
    return response.data;
  },

  typeList: async () => {
    const response = await api.post<ResultDTO<Array<{ key: string; name: string }>>>(
      `${BASE}/type-list`,
    );
    return response.data;
  },
};
